pkg/controller/master/backup/backup.go

```diff
@@ -180,31 +180,57 @@ func (h *Handler) getVolumeBackups(backup *harvesterv1.VirtualMachineBackup, vm
 
 // getSecretBackups helps to build a list of SecretBackup upon the cloud init secrets used by the backup VM
 func (h *Handler) getSecretBackups(vm *kv1.VirtualMachine) ([]harvesterv1.SecretBackup, error) {
-	var secretBackups []harvesterv1.SecretBackup
+	secretBackupMap := map[string]harvesterv1.SecretBackup{}
 
 	for _, volume := range vm.Spec.Template.Spec.Volumes {
 		if volume.CloudInitNoCloud != nil && volume.CloudInitNoCloud.UserDataSecretRef != nil {
 			secret, err := h.secretCache.Get(vm.Namespace, volume.CloudInitNoCloud.UserDataSecretRef.Name)
 			if err != nil {
 				return nil, err
 			}
-			secretBackups = append(secretBackups, harvesterv1.SecretBackup{
+
+			data := secret.DeepCopy().Data
+			for k, v := range secret.Data {
+				if len(v) == 0 {
+					delete(data, k)
+				}
+			}
+
+			secretFullName := fmt.Sprintf("%s/%s", secret.Namespace, secret.Name)
+			secretBackupMap[secretFullName] = harvesterv1.SecretBackup{
 				Name: secret.Name,
-				Data: secret.Data,
-			})
+				Data: data,
+			}
 		}
 		if volume.CloudInitNoCloud != nil && volume.CloudInitNoCloud.NetworkDataSecretRef != nil {
 			secret, err := h.secretCache.Get(vm.Namespace, volume.CloudInitNoCloud.NetworkDataSecretRef.Name)
 			if err != nil {
 				return nil, err
 			}
-			secretBackups = append(secretBackups, harvesterv1.SecretBackup{
+			secretFullName := fmt.Sprintf("%s/%s", secret.Namespace, secret.Name)
+			// users may put UserDataSecretRef and NetworkDataSecretRef in a same secret, so we only keep one
+			if _, ok := secretBackupMap[secretFullName]; ok {
+				continue
+			}
+
+			data := secret.DeepCopy().Data
+			for k, v := range secret.Data {
+				if len(v) == 0 {
+					delete(data, k)
+				}
+			}
+
+			secretBackupMap[secretFullName] = harvesterv1.SecretBackup{
 				Name: secret.Name,
-				Data: secret.Data,
-			})
+				Data: data,
+			}
 		}
 	}
 
+	secretBackups := make([]harvesterv1.SecretBackup, 0, len(secretBackupMap))
```

Comment:
It looks the secretBackupMap is used to skip duplicated names, when yes,  we may consider to improve it:

`secretBackupMap := map[string]harvesterv1.SecretBackup{} `  -----> `map[string]int/bool`

in the loop,  
when volume.CloudInitNoCloud.UserDataSecretRef
  generate name, add name into map, add Data into  secretBackups 
 volume.CloudInitNoCloud.NetworkDataSecretRef
   generate name, check if in map: yes, continue;  no, add name into map, add Data into secretBackups 

finally, return `secretBackups `directly, no need to do following for loop  to generate it

Reply:
Good catch. Updated it. Thank you.